rustler_bigint/src/big_int.rs

```
@@ -0,0 +1,396 @@
+        97 => {
+            // small integer / byte
+            num_bigint::BigInt::from(input[2])
+        }
+
+        98 => {
+            // integer
+            num_bigint::BigInt::from_signed_bytes_be(&input[2..6])
+        }
+
+        110 => {
+            // small big integer
+            let n = input[2] as usize;
+            let sign = if input[3] == 0 {
+                Sign::Plus
+            } else {
+                Sign::Minus
+            };
+
+            num_bigint::BigInt::from_bytes_le(sign, &input[4..n + 4])
+        }
+
+        111 => {
+            // large big integer
+            let n = u32::from_be_bytes([input[2], input[3], input[4], input[5]]) as usize;
+            let sign = if input[6] == 0 {
+                Sign::Plus
+            } else {
+                Sign::Minus
+            };
+
+            num_bigint::BigInt::from_bytes_le(sign, &input[7..n + 7])
+        }
+
+        _ => return Err(Error::BadArg),
+    };
+
+    Ok(big_int)
+}
+
+fn encode_big_integer(big_int: &num_bigint::BigInt) -> Vec<u8> {
+    let (sign, data) = big_int.to_bytes_le();
+    let sign = if sign == Sign::Minus { 1 } else { 0 };
+
+    let mut out = vec![131];
+    if data.len() < 256 {
+        // small big integer
+        let n = data.len() as u8;
+        out.push(110);
+        out.push(n);
+    } else {
+        // large big integer
+        let n = (data.len() as u32).to_be_bytes();
+        out.push(111);
+        out.extend(n);
+    };
+    out.push(sign);
+    out.extend(data);
+
+    out
+}
+
+impl<'a> Decoder<'a> for BigInt {
+    fn decode(term: Term<'a>) -> NifResult<Self> {
+        decode_big_integer(term.to_binary().as_slice()).map(BigInt)
+    }
+}
+
+impl Encoder for BigInt {
+    fn encode<'c>(&self, env: Env<'c>) -> Term<'c> {
+        let binary = encode_big_integer(&self.0);
+        if let Some((term, _)) = env.binary_to_term(&binary) {
+            term
+        } else {
+            error().encode(env)
```

Do you know how this error could be triggered?

---

I think we can/should assume that `encode_big_integer` is correct. Also, should we maybe call `encode(env)` on the `Vec<u8>`?

```rust
fn encode<'c>(&self, env: Env<'c>) -> Term<'c> {
  encode_big_integer(&self.0).encode(env)
}
```

---

We have to still convert it from etf bytes to a valid erlang/elixir term. 
Did a small refactor to return nil instead when invalid bytes are encountered (ofcourse it should never happen)

---

Maybe we should raise an exception if conversion fails, e.g. `::rustler::Error::RaiseTerm(format!("could not encode bigint {}", bigint))`. What do you think? I think `nil` might be quite unexpected.

---

Yes, lets do that. Found the nil thing also quite akward while I implemented it :sweat_smile:

---

Small issue, is it possible to convert an Error to a Term? I don't think we can right? (because how can you encode a raise statement in a term :thinking: )

---

updated to return an error tuple

---

Looks good, thanks!